gdal/ogr/ogrsf_frmts/ngw/ngw_api.cpp

```diff
@@ -0,0 +1,643 @@
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *******************************************************************************/
+
+#include "ogr_ngw.h"
+
+#include "cpl_http.h"
+
+namespace NGWAPI {
+
+std::string GetPermisions(const std::string &osUrl, const std::string &osResourceId)
+{
+    return osUrl + "/api/resource/" + osResourceId + "/permission";
+}
+
+std::string GetResource(const std::string &osUrl, const std::string &osResourceId)
+{
+    return osUrl + "/api/resource/" + osResourceId;
+}
+
+std::string GetChildren(const std::string &osUrl, const std::string &osResourceId)
+{
+    return osUrl + "/resource/" + osResourceId + "/child/";
+}
+
+std::string GetFeature(const std::string &osUrl, const std::string &osResourceId)
+{
+    return osUrl + "/api/resource/" + osResourceId + "/feature/";
+}
+
+std::string GetTMS(const std::string &osUrl, const std::string &osResourceId)
+{
+    return osUrl + "/api/component/render/tile?z=${z}&amp;x=${x}&amp;y=${y}&amp;resource=" +
+        osResourceId;
+}
+
+std::string GetFeaturePage(CPL_UNUSED const std::string &osUrl, CPL_UNUSED const std::string &osResourceId,
+    CPL_UNUSED GIntBig nStart, CPL_UNUSED int nCount)
+{
+    return ""; // TODO: Add paging support.
+}
+
+std::string GetRoute(const std::string &osUrl)
+{
+    return osUrl + "/api/component/pyramid/route";
+}
+
+Uri ParseUri(const std::string &osUrl)
+{
+    Uri stOut;
+    std::size_t nFound = osUrl.find(":");
+    if( nFound == std::string::npos )
+    {
+        return stOut;
+    }
+
+    stOut.osPrefix = osUrl.substr(0, nFound);
+    std::string osUrlInt = CPLString(osUrl.substr(nFound + 1)).tolower();
+
+    nFound = osUrlInt.find("/resource/");
+    if( nFound == std::string::npos )
+    {
+        return stOut;
+    }
+
+    stOut.osAddress = osUrlInt.substr(0, nFound);
+
+
+
+    std::string osResourceId = CPLString(osUrlInt.substr(nFound + 10)).Trim();
+
+    nFound = osResourceId.find('/');
+    if( nFound != std::string::npos )
+    {
+        stOut.osResourceId = osResourceId.substr(0, nFound);
+        stOut.osNewResourceName = osResourceId.substr(nFound + 1);
+    }
+    else
+    {
+        stOut.osResourceId = osResourceId;
+    }
+
+    return stOut;
+}
+
+static void ReportError(const GByte *pabyData, int nDataLen)
+{
+    CPLJSONDocument oResult;
+    if( oResult.LoadMemory(pabyData, nDataLen) )
+    {
+        CPLJSONObject oRoot = oResult.GetRoot();
+        if( oRoot.IsValid() )
+        {
+            std::string osErrorMessage = oRoot.GetString("message");
+            if( !osErrorMessage.empty() )
+            {
+                CPLErrorSetState(CE_Failure, CPLE_AppDefined, osErrorMessage.c_str());
```

Comment:
Didn't you mean CPLError() instead ? CPLErrorSetState() just states the error state, but doesn't trigger error callbacks

Reply:
Actually I see there are several other calls to CPLErrorSetState()

Reply:
Fix per 4579aeab927d70ddab6142340bf5550c6789b42e